### PR TITLE
Docker fixes

### PR DESCRIPTION
**Pull request by dbnicholson**
Here are various fixes I put together when working with the build in a Docker container. I believe everything should work the same as before with the documented `build_docker` and `run_docker` make targets.

### CODE REVIEW

File: Dockerfile
Comment (DXCanas) [inline Dockerfile:60]: Again, acknowledging GH commit messages:
https://github.com/learningequality/kolibri-installer-android/pull/107/commits/a74373b6cd811f58519762afcac5fb5e143755f4
https://github.com/learningequality/kolibri-installer-android/pull/107/commits/edcd28b85034730d1041ead505ec9654147fa5d6

I don't see an issue with the `ndk-bundle` softlink, but I'll defer to @rtibbles

I do want to point out a downside of including these definitions here: multiple sources of truth. If there need to be any changes to the envars there are now 2 places to update, which defeats the spirit of going through Makefile (not saying that we're flawless in this regard now though)

File: scripts/rundocker.sh
Comment (DXCanas) [inline scripts/rundocker.sh:32]: https://github.com/learningequality/kolibri-installer-android/pull/107/commits/63010a261e677e9872a5a09b37ef5dbcd8a325e7
> Without this, gradlew downloads gradle every time the container is run.

🙌

**Review by DXCanas — CHANGES_REQUESTED** [review]

**Comment by DXCanas** [comment]
Submitted so quickly I forgot to add a summary to my review:
Overall, these are all really neat changes. 

Thanks a lot for the thought put in here. Deferred to others for some, and requested changes based primarily on the cache busting concerns and "multiple sources of truth" scenario.

**Comment by DXCanas** [comment]
Submitted so quickly I forgot to add a summary to my review:
Overall, these are all really neat changes. 

Thanks a lot for the thought put in here. Deferred to others for some, and requested changes based primarily on the cache busting concerns and "multiple sources of truth" scenario.

File: scripts/rundocker.sh
Comment (dbnicholson) [reply scripts/rundocker.sh:26]: Sure, I'll add a couple comments.

File: scripts/rundocker.sh
Comment (dbnicholson) [reply scripts/rundocker.sh:41]: I haven't totally settled on the workflow I really like, so the `ENTRYPOINT` one could probably be dropped off for now. But the biggest reason I'm doing this now is to build the unsigned debug APK. Without passing this through, you always run `make kolibri.apk` from this script. Also, right now I'm mostly using the Makefile targets, but if I want to pass any different arguments to `p4a` then I need to get those arguments through docker and make. On the other hand, if I can call the command directly, I can just run `p4a` with the exact arguments I want without having to marshal them through the Makefile.

But generally I like that I can get into the environment with the cache mounted in the right place and look around. For example, I've been running `./scripts/rundocker.sh bash` and then running `p4a` directly so I can see what's going on a little more closely.

File: Dockerfile
Comment (dbnicholson) [reply Dockerfile:81]: At least on my laptop, the console goes all over the place when gradle uses it's default `auto` setting. I suspect this is really a bug in gradle and it's just checking if the console is a TTY when it likely needs to check the capabilities of the TTY in some way.

The daemon probably doesn't make a difference one way or the other. It certainly doesn't help you if the entire procedure is to start a container, run it once and kill it. But I guess it's probably very little overhead to wait for it to start and then have the client communicate with it.

File: Dockerfile
Comment (dbnicholson) [reply Dockerfile:64]: I think the `ARCH` environment variable definitely does have use. All I'm saying is that in the Dockerfile it doesn't actually do anything. If you wanted to make it work there, you'd have to actually use `--build-arg` somewhere. docker build doesn't use environment variables. I tried both of the following and neither did anything when I had ARCH set as an environment variable:

```
FROM debian:stable-slim
ARG ARCH=${ARCH}
RUN echo ARCH=$ARCH
```

```
FROM debian:stable-slim
ENV ARCH=${ARCH}
ARG ARCH=${ARCH}
RUN echo ARCH=$ARCH
```

Only when I used `--build-arg` or changed to `ENV ARCH=something` did it actually echo anything.

File: Dockerfile
Comment (dbnicholson) [reply Dockerfile:69]: Yes, the cache invalidation on unrelated `Makefile` changes is really unfortunate. The only real advantage I guess is that (for example) changing an asset doesn't cause you to reinstall the SDK.

My other thought here was to split out the SDK install commands to a script instead of having them embedded in the Makefile. Right now no matter where you put it in the Dockerfile, you need the Makefile in the build container before you can install the SDK. Whether that's by root or unprivileged and whether it's just the Makefile or the entire checkout being copied, you're going to reinstall the SDK. And that's really slow.

I haven't seen any issue using the SDK as installed by root. That's what happens in the Ubuntu VMs used in the Github workflows. They're not installing the SDK at all and are instead using the pre-installed one in /usr/local/lib/android.

File: Dockerfile
Comment (dbnicholson) [reply Dockerfile:60]: I kind of see it the other way. When you run the Makefile normally, it requires that you have `ANDROID_HOME`, `ANDROIDSDK` and `ANDROIDNDK` setup in the environment already. Having those environment variables embedded in the image means that's already taken care of in the container. It makes the image very analogous to the [Ubuntu Github runners](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md#environment-variables-3).

File: scripts/rundocker.sh
Comment (dbnicholson) [reply scripts/rundocker.sh:32]: I did get really tired of watching that get downloaded.

I think this could actually be the entire `~/.gradle` directory without an issue. Before I figured out the `GRADLE_OPTS` environment variable, I was messing with `~/.gradle/gradle.properties`. Then I had a conflict with managing the volume vs pre-seeding part of it.

File: Dockerfile
Comment (rtibbles) [reply Dockerfile:64]: One thing to flag is that once we upgrade to the latest version of Python-for-Android, it now has multi-arch support, so we may not need to do more than one build, if I am understanding it correctly.

File: Dockerfile
Comment (rtibbles) [reply Dockerfile:69]: I don't think there's any issue with running as root inside the container, as @dbnicholson says, it is what happens inside Github Actions containers.

I had originally put as much as possible into the Makefile, in order to allow for easy invalidation of caches when the Makefile changes. It's a slightly dumb mechanism, but I've had more issues with cache not invalidating than with it invalidating too often, so I've pushed it the other way in my previous work here.

File: Dockerfile
Comment (rtibbles) [reply Dockerfile:60]: The only reason this stuff isn't _only_ in the Dockerfile, but also referenced in the Makefile is to allow local devs to leverage the same tooling outside of a container, I think a little bit of repetition here is fine.

**Comment by dbnicholson** [comment]
I started rebasing this on #110 and while poking at this again and found I could vastly improve this setup so long as the p4a in use has kivy/python-for-android#2582. Then you don't need the `kivy` user in the container at all. I made a PR in learningequality/python-for-android#1 to do that.

File: Dockerfile
Comment (dbnicholson) [reply Dockerfile:64]: I skimmed past the earlier comment about fear of using the same image for 64 and 32 bit builds. I actually think it's fine to use the same image for both. Nothing in the image depends on the target architecture. I also think it would be fine to share the cache between architectures. p4a separates the build outputs by target architecture. But maybe I missed something there.

And like @rtibbles says, with the current version of p4a you can do a multi-arch build.

Comment (rtibbles) [reply Dockerfile:64]: Except that they haven't enabled it for the webview bootstrap that we use yet, so we'll probably have to do that port on our fork, and contribute it back upstream.

File: Dockerfile
Comment (DXCanas) [reply Dockerfile:81]: No reason to get rid of it, just curious!

File: Dockerfile
Comment (DXCanas) [reply Dockerfile:69]: Personally, I like that most of the logic is in the Makefile - provides a unified location to update scripts for both docker users and local builders. Puts in in a place where either can modify it and propagate changes to the other workflow. 

Which was my main "argument" against the `ENV` [definitions below](https://github.com/learningequality/kolibri-installer-android/pull/107#discussion_r852413114). 

Given that the cache would be invalidated with Makefile changes, I was questioning whether the layer rearrangement was worth it; I figure a change to Makefile is high on the "most likely to be updated" list. But your point about changing an asset is valid. 

I'm not doing active dev here, but my impression is that this repo isn't meant to change very often at all, regularly building based off of updated Kolibri artifacts. Right now, those are copied in during build time. And moving `setup` back is probably the right move. Let's keep it! 

I'm unsure what the intention behind the proceeding `rm`s are, but I don't see any reason to remove them 🤷

File: Dockerfile
Comment (DXCanas) [reply Dockerfile:60]: If @rtibbles is cool with it, I have no objections!

**Review by DXCanas — APPROVED** [review]
All the functional changes I requested had are dismissed!

File: scripts/rundocker.sh
Comment (dbnicholson) [reply scripts/rundocker.sh]: I ended up dropping this commit when I was able to do this all with a single `docker run` command rather than starting and stopping the container.

**Comment by dbnicholson** [comment]
Heh, you merged so fast I didn't even have time to describe something I changed :)

The last commit is the real reason I set off to do all of this - run entirely as the current user without any copying in or out of the image and without embedding the source into the image. In my testing it works great, but that's not what got reviewed before. I'm happy to switch that back, but I do think it's better and more robust this way.

File: scripts/rundocker.sh
Comment (DXCanas) [reply scripts/rundocker.sh]: I actually missed that change.

No complaints here about no longer needing the `kivy` user as long as permissions hold up, [which sounds to be the case ](https://github.com/learningequality/kolibri-installer-android/commit/31f33d0119c2f55df8ad0ed53d859386b230aa3b)🤷 Thanks!!

Glad we can track upstream more closely -- seems to be paying off in lots of ways @rtibbles

**Comment by rtibbles** [comment]
I read the code (quickly) before I merged - and those changes made sense. I think given that P4A does have this implicit dependency on the user HOME dir, it makes a lot of sense to properly alias that, rather than trying to workaround it by moving stuff in and out of the container. Not embedding the source into the image also seems preferable.

If this somehow breaks something later, then I'll complain!

**Pull request merged**
